src/next.ts

```diff
@@ -0,0 +1,66 @@
+import areInputsEqual from './are-inputs-equal';
+
+// Using ReadonlyArray<T> rather than readonly T as it works with TS v3
+export type EqualityFn<TFunc extends (...args: any[]) => any> = (
+  newArgs: Parameters<TFunc>,
+  lastArgs: Parameters<TFunc>,
+) => boolean;
+
+type Cache<TFunc extends (this: any, ...args: any[]) => any> = {
+  lastThis: ThisParameterType<TFunc>;
+  lastArgs: Parameters<TFunc>;
+  lastResult: ReturnType<TFunc>;
+};
+
+type MemoizedFn<TFunc extends (this: any, ...args: any[]) => any> = {
```

Comment:
I would propably do
```ts
type MemoizedFn<TFunc extends (this: any, ...args: any[]) => any> = TFunc & {
  clear: () => void;
};
```

Reply:
That's an open question - should `.someProps` be dropped or preserved.

Reply:
Opinion: I think additional function props should be dropped. This is what `memoize-one` has always done.

Reply:
The only thing I would change is using `interface` instead of `type`. `interface`s are [the recommended way](https://github.com/microsoft/TypeScript/wiki/Performance#preferring-interfaces-over-intersections), since it's faster (related to TS compiler)

Reply:
Thanks for all your input @oriSomething + @theKashey ❤️